### PR TITLE
nixos/doc: render option values using `lib.generators.toPretty`

### LINKED ISSUE #195562: Examples in NixOS options manual do not escape attribute names
## Problem

Attribute sets in the manual's examples do not properly escape keys when are, for example, paths.  For the "Example" for an option, we see `{ /home/user = "foo" }` which is invalid Nix code instead of `{ "/home/user" = "foo" }` which is valid Nix code.  This happens when the Nix options get `builtins.toJSON` run on them during the manual generation.  In the wild, look at `/mnt/btr_pool` in the example of [this option](https://nixos.org/manual/nixos/stable/options.html#opt-services.btrbk.instances._name_.settings).  Maybe the general opinion on this is, "obviously attrset keys can't be paths", but it tripped me up for a few minutes as a newcomer, and I think it could be a nice quality of life improvement.

## Checklist

- [x] checked [latest Nixpkgs manual] \([source][nixpkgs-source]) and [latest NixOS manual] \([source][nixos-source])
- [x] checked [open documentation issues] for possible duplicates
- [x] checked [open documentation pull requests] for possible solutions

[latest Nixpkgs manual]: https://nixos.org/manual/nixpkgs/unstable/
[latest NixOS manual]: https://nixos.org/manual/nixos/unstable/
[nixpkgs-source]: https://github.com/NixOS/nixpkgs/tree/master/doc
[nixos-source]: https://github.com/NixOS/nixpkgs/tree/master/nixos/doc/manual
[open documentation issues]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+label%3A%229.needs%3A+documentation%22
[open documentation pull requests]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+is%3Apr+label%3A%228.has%3A+documentation%22%2C%226.topic%3A+documentation%22

## Proposal

Changing `toJSON` in Nix seems inappropriate, so maybe I could open a PR with a Nix function which escapes the problem attributes, but that might have performance issues.  I am open to suggestions (or just marking it as "won't fix").

### PR DESCRIPTION
**Pull request by ncfavier**
Fixes https://github.com/NixOS/nixpkgs/issues/195562, see discussion there.

A nice side effect is that we can leverage `toPretty`'s `allowPrettyValues` feature to make scrubbed derivations pretty-print as `pkgs.foo`. This should allow us to *mostly* get rid of the

```
default = pkgs.foo;
defaultText = lib.literalExpression "pkgs.foo";
```

pattern which is so easy to get wrong (though instances like `cfg.package` would still require it).

Note that if this approach works out, we can maybe turn it into a *promise* that `optionAttrSetToDocList` does not output non-rendered Nix values, so that we can eventually get rid of all the other Nix rendering code, in-tree and out-of-tree.

Regarding the first commit, I am aware that most of these issues would be avoided by adding proper type checks to defaults and examples. I'll try to look into this next.

### CODE REVIEW

File: nixos/modules/services/matrix/synapse.nix
Comment (ncfavier) [inline nixos/modules/services/matrix/synapse.nix:289]: I'm not sure what should be done with this. Maybe `toPretty` should somehow be made aware of `pkgs.path` so that it can render internal paths independently of where nixpkgs is located?

Comment (roberth) [reply nixos/modules/services/matrix/synapse.nix:289]: The documentation generator already needs to deal with paths for the declaration and definition locations. The logic for this is a bit strewn about. It'd make sense to make the generic doc generation take ownership of that functionality and then reuse that logic by passing it to `toPretty`.

**Review by pennae — APPROVED** [review]
looks reasonable, but we'd still rather not convert everything to text in nix. the generated manual does look a lot better on composite defaults though, so that's a big plus. might be worth it to make the average docs better, even if more advanced systems would have to re-parse examples.

> This should allow us to mostly get rid of the [defaultText for packages]

not really, since `pkgs` is a throw thunk in the lazy module docs build. might be better to use `mkPackageOption` instead for options referring to packages since that takes care of the duplication.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (ncfavier) [reply nixos/modules/virtualisation/qemu-vm.nix:598]: I think you meant `"config.${options.virtualisation.host.pkgs}.qemu_vm"`, but that fails with `attribute 'host' missing` for some reason.

**Review by roberth — APPROVED** [review]

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (ncfavier) [reply nixos/modules/virtualisation/qemu-vm.nix:598]: Yeah that doesn't work either because `virtualisation` is not an option.

The other error comes from the fact that we [override](https://github.com/NixOS/nixpkgs/blob/6a36558f5227f728d4fa433ab8ec1b80bebe4f50/nixos/modules/misc/documentation.nix#L43-L45) the `options` argument with the outer `options` so that eager modules have access to all options, but indeed "all" doesn't include the `qemu-vm.nix` module by default.

I looked into doing something like `options = lib.recursiveUpdate allOpts eagerOpts`, where `eagerOpts = scrubbedEval.options`, but this runs into infinite recursion because those are the *`checked`* options. We would need to be able to access the unchecked `options` fixed point, which could be done either by adding `uncheckedOptions = options;` [here](https://github.com/NixOS/nixpkgs/blob/6a36558f5227f728d4fa433ab8ec1b80bebe4f50/lib/modules.nix#L328-L333) or by allowing `specialArgs` to be a function that accepts `options`.

In any case this is way beyond the scope of this PR.

**Comment by ncfavier** [comment]
> more advanced systems would have to re-parse examples.

What do you have in mind?

File: nixos/modules/misc/documentation.nix
Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:58]: I think I agree. My main concern are things like `default = "... ${pkgs.foo} ..."` without a `defaultText`. The `"\${${wholeName}}"` hack currently makes this render "nicely" but this is arguably a bug with `toPretty`: the interpolation should be escaped, so that it doesn't look nice anymore. I think I'd like to do something like

```
// optionalAttrs (isDerivation value) {
  outPath = throw "attempt to use a derivation in options documentation";
  drvPath = # same
}
```

It's a bit annoying that this has to be decided at the level of the evaluation, rather than in the generic `make-options-doc`, so that out-of-tree projects have to make their own decision. But probably fine.

Comment (roberth) [reply nixos/modules/misc/documentation.nix:58]: > The `"\${${wholeName}}"` hack currently makes this render "nicely" but this is arguably a bug with `toPretty`:

I agree that this is a bug, but is it a reason to drop easy `pkgs.foo` interpolation from the docs?

It could be made to work well by
 - fixing the escaping first
 - making the escaping customizable (function passed to `toPretty`)
 - pass a custom escaping function, which at the end replaces a special syntax like `$$${{{` to `${` so that antiquotations are representable without stealing too much from the normal string literals.
 - make `pkgs.foo.outPath = "\$\$\${{{pkgs.foo}"`
 - toPretty will print the right expression

Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:58]: Considering that this feature is not used by anything in-tree (manual still builds with a `throw`), I think it's worth trying to move in the direction of least hackiness, which IMO is something like https://github.com/NixOS/nix/issues/1474.

Comment (roberth) [reply nixos/modules/misc/documentation.nix:58]: Alright. Then let's keep using the least hacky option available with the Nix we have today:

```nix
{
  defaultText = lib.literalExpression ''
    "''${pkgs.foo}/bin/foo"
  '';
}
```

**Comment by ncfavier** [comment]
Alright, I got a bit carried away... see commits.

### Notes

- This PR changes the format of `options.json` (declarations can now be {path, url} instead of strings), which breaks nixos-search. This should be fixed *before* merging.
  Users of `optionAttrSetToDocList` should not get too broken (e.g. the home-manager manual still looks fine).
- After merging, the [release wiki](https://github.com/NixOS/release-wiki) should be updated to mention `lib.trivial.manualRevision`.
- After merging, morph can undo most of https://github.com/DBCDK/morph/pull/108/files and instead set `documentation.nixos.extraModuleSources = [ { name = "morph"; prefix = ../.; urlFor = m: ...; }]`. cc @bb010g
- https://github.com/NixOS/nix/pull/6530#issuecomment-1309963199 will break `lazy-options.json` because we now filter the nixpkgs source based on relative paths.
- An alternative to the `moduleSources` argument would be to add a `_module.sources` option so that modules can define their own sources. This could work either by also passing `config` to `optionsToDocTemplate`, or by making a `modulesToDocTemplate` wrapper which calls `evalModules`. I haven't thought this through, just throwing ideas around.

**Review by pennae — COMMENTED** [review]
looks mostly reasonable. rendered options seem to fine judging by the limited number of checks we've done.

File: nixos/modules/misc/documentation.nix
Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:76]: The `.nix` filtering never worked, because the check was `t == "file"` instead of `t == "regular"`, and in fact excluding non-nix files isn't possible currently because of e.g. https://github.com/NixOS/nixpkgs/blob/cecc553095ea244e516003f7e0714669afb4af70/nixos/modules/services/system/nscd.nix#L57.

We could still make it work by adding a `defaultText` there, or marking that module non-cacheable, or we could just filter out directories named `doc`

File: nixos/modules/misc/documentation.nix
Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:306]: The problem is compatibility: previously `extraModuleSources` only accepted paths to be stripped, so the question is what to do with those. Currently they are mapped to a nameless source, I don't see a better alternative short of throwing an error, but that seems a bit brutal.

File: lib/generators.nix
Comment (ncfavier) [reply lib/generators.nix:297]: I guess *semantically* this should be a mapping from names to paths, but *technically* it should be a mapping from paths to names (because that's the direction of the transformation we're doing), so I went the neutral route and picked neither.

I'm fine with going the semantic route though.

File: nixos/modules/misc/documentation.nix
Comment (bb010g) [reply nixos/modules/misc/documentation.nix:306]: `<_unknown/*>` may be a good default?

Comment (roberth) [reply nixos/modules/misc/documentation.nix:306]: Perhaps `baseNameOf config.prefix` and maybe strip off a store path hash part, if there is one.
I guess that will effectively be `"source"` in many cases.

What do the angle brackets, underscore, slash and asterisk mean? It seems a bit excessive.

Comment (bb010g) [reply nixos/modules/misc/documentation.nix:306]: @roberth Imagine a NixOS module source under version control, checked out in a home directory. If there's a module at the top level, e.g. `$CHECKOUT/configuration.nix`, the base name of `$CHECKOUT` isn't necessarily stable between checkouts. We should avoid looking at parents of the top-level. If we don't provide a default module source name, this path renders in documentation as `<configuration.nix>`. If the source then imports a module from `$CHECKOUT/nixpkgs/nixos/system/build.nix`, its path then renders as `<nixpkgs/nixos/system/build.nix>`, which is the path of an existing Nixpkgs module. (This example is contrived, but the situation can reasonably occur if multiple external module sets use simple paths.) On the other hand, if we provide a default module source name of `_unknown/`, then the former path renders as `<_unknown/configuration.nix>` and the latter path renders as `<_unknown/nixpkgs/nixos/system/build.nix>`.

This approach can still result in collisions, but a default name of `_unknown` sticks out and looks ugly enough that anyone who notices will change it, and the leading underscore is conventional for internal Nix & Nixpkgs names.

If we wanted the name to stick out more and be easier to search, `_unknown-module-source` could work really well (`<_unknown-module-source/configuration.nix>`).

On the `…/configuration.nix` note from earlier, we may want to add `config.nixos.extraModuleSources = [ { name = "nixos-config"; prefix = ./.; } ];` to the `/etc/nixos/configuration.nix` template.

Comment (roberth) [reply nixos/modules/misc/documentation.nix:306]: Ok, now I understand what you meant.

I'd be quite happy for it to throw an exception in projects where I control the scope of docs generation actually, but for NixOS, returning something ugly seems like a pretty good, robust solution.

We can't really guess the name, and we certainly can't let the NixOS evaluation fail because of the docs, so `<_unknown/whatever/path/file.nix>` seems like a good solution.

Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:306]: Agreed, will do this.

> we may want to add `config.nixos.extraModuleSources = [ { name = "nixos-config"; prefix = ./.; } ];` to the `/etc/nixos/configuration.nix` template.

I'm not convinced. Seems like a lot of boilerplate for something the vast majority of people won't care about (and if that file is indeed destined to be in `/etc/nixos/configuration.nix`, then `<nixos-config/X>` isn't a lot better than `/etc/nixos/X`.

Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:306]: Actually, here's an alternative: introduce a new `moduleSources` option with the "ideal" type (attrsOf {prefix, urlFor}), and add a `mkChangedOptionModule` that maps paths from `extraModuleSources` to attributes in `moduleSources` using the mechanism proposed by @roberth (basename of prefix minus store hash). Seems like the best of all worlds?

Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:306]: I guess this will break if someone has two paths named `source`, which is quite likely...

Comment (roberth) [reply nixos/modules/misc/documentation.nix:306]: I thought `name` was only a label for presentation, not a lookup key. Isn't prefix the only lookup key?

Comment (roberth) [reply nixos/modules/misc/documentation.nix:306]: Lookup key in the logical sense. I think prefix has to be a linear search, not an attrset key, because it can be a path value, and we may not want to `toString` those because of lazy trees.

Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:306]: See the mirror [discussion](https://github.com/NixOS/nixpkgs/pull/199363#discussion_r1021733671) for `searchPath`. Technically you're right, but I think it makes sense to have module sources indexed by their name. We might find more use cases for them later than just stripping prefixes.

Comment (roberth) [reply nixos/modules/misc/documentation.nix:306]: You could use `name` as an extra key that's an actual key and have a `displayName` option for the cases where we don't have enough info.

Comment (ncfavier) [reply nixos/modules/misc/documentation.nix:306]: Not sure I understand the use case for that, and in any case it seems well beyond the scope of this PR so I think I'll revert to the "least disruptive" choice (the current state of this PR, but make `name` non-optional and default to `_unknown` for raw paths) and let others extend this later.

**Comment by ncfavier** [comment]
I went with my `_module.sources` idea and isolated the prefix lookup logic into a new function `lib.strings.lookupPrefix`.
That way it's not NixOS-specific, module sources are responsible for identifying themselves, and we can bake the nixpkgs source in (which will always be used for the `_module` options if those are rendered).

Also moved the `<source/path>` rendering down to XSLT; other consumers get a {source, path, url?} attrset which they are free to render as they wish.

I still think it's fine to allow sources with colliding `name`s, e.g. when Nix gets lazy trees we might somehow want to have roots for both `"${flake}"` and `toString flake` in certain situations.

I've tested this in a variety of situations and I've got PRs ready for home-manager, nmd and nixos-search.

File: nixos/modules/misc/documentation.nix
Comment (ncfavier) [reply nixos/modules/misc/documentation.nix]: I think we should avoid bothering module authors with this implementation detail as much as possible, so I went with excluding `doc`. The resulting store path weighs 13.5 MiB, of which 12.6 MiB are .nix files.

**Comment by fricklerhandwerk** [comment]
Oof, that's a large one. I suppose I was added automatically based on `CODEOWNERS`. Sorry @ncfavier, I won't have time to review this any time soon, maybe never, but you seem to have plenty of reviewers who can help merge this.

**Comment by infinisil** [comment]
This PR does too many things now! Can you split this up into smaller reviewable bits @ncfavier?

**Comment by ncfavier** [comment]
Indeed, forked the path stuff into https://github.com/NixOS/nixpkgs/pull/203994.

Note that the commits left in this PR are already reviewed (but additional reviews can't hurt!)

File: nixos/lib/make-options-doc/options-to-docbook.xsl
Comment (ncfavier) [inline nixos/lib/make-options-doc/options-to-docbook.xsl:197]: Note there's already a template for `literalExpression` above.

**Comment by ncfavier** [comment]
I think this is good to merge now?

**Review by pennae — COMMENTED** [review]
yep, looks good (except for a few nitpicks. just ignore them if you want to get this over with.)

File: nixos/modules/services/databases/openldap.nix
Comment (ncfavier) [reply nixos/modules/services/databases/openldap.nix:19]: I know, but raw MarkDown is at least more readable than raw DocBook.

File: lib/generators.nix
Comment (ncfavier) [reply lib/generators.nix]: Alright, I went with `toJSON` and left a comment

**Pull request merged**
